loader/loader.c

```diff
@@ -767,10 +768,241 @@ VkResult loader_add_to_dev_ext_list(const struct loader_instance *inst, struct l
-                           const struct loader_layer_list *source_list);
+// Determine the type of filter string based on the contents of it.
+// This will properly check against:
+//  - substrings "*string*"
+//  - prefixes "string*"
+//  - suffixes "*string"
+//  - full string names "string"
+// It will also return the correct start and finish to remove any star '*' characters for the actual string compare
+void determine_filter_type(const char *filter_string, enum loader_filter_string_type *filter_type, const char **new_start,
+                           size_t *new_length) {
+    size_t filter_length = strlen(filter_string);
+    bool star_begin = false;
+    bool star_end = false;
+    if ('~' == filter_string[0]) {
+        // One of the special identifiers like: ~all~, ~implicit~, or ~explicit~
+        *filter_type = FILTER_STRING_SPECIAL;
+        *new_start = filter_string;
+        *new_length = filter_length;
+    } else {
+        if ('*' == filter_string[0]) {
+            // Only the * means everything
+            if (filter_length == 1) {
+                *filter_type = FILTER_STRING_SPECIAL;
+                *new_start = filter_string;
+                *new_length = filter_length;
+            } else {
+                star_begin = true;
+            }
+        }
+        if ('*' == filter_string[filter_length - 1]) {
+            // Not really valid, but just catch this case so if someone accidentally types "**" it will also mean everything
+            if (filter_length == 2) {
+                *filter_type = FILTER_STRING_SPECIAL;
+                *new_start = filter_string;
+                *new_length = filter_length;
+            } else {
+                star_end = true;
+            }
+        }
+        if (star_begin && star_end) {
+            *filter_type = FILTER_STRING_SUBSTRING;
+            *new_start = &filter_string[1];
+            *new_length = filter_length - 2;
+        } else if (star_begin) {
+            *new_start = &filter_string[1];
+            *new_length = filter_length - 1;
+            *filter_type = FILTER_STRING_SUFFIX;
+        } else if (star_end) {
+            *filter_type = FILTER_STRING_PREFIX;
+            *new_start = filter_string;
+            *new_length = filter_length - 1;
+        } else {
+            *filter_type = FILTER_STRING_FULLNAME;
+            *new_start = filter_string;
+            *new_length = filter_length;
+        }
+    }
+}
+
+// Parse the provided filter string provided by the envrionment variable into the appropriate filter
+// struct variable.
+void parse_generic_filter_environment_var(const struct loader_instance *inst, const char *env_var_name,
+                                          struct loader_envvar_filter *filter_struct) {
+    memset(filter_struct, 0, sizeof(struct loader_envvar_filter));
+    char *env_var_value = loader_secure_getenv(env_var_name, inst);
+    if (NULL != env_var_value) {
+        if (strlen(env_var_value) > 0) {
```

Comment:
Style wise, its better to do `if (NULL == env_var_value ) return;`, so the success path isn't as indented.

Reply:
👍🏽